Validate nested rulesets

simonsmith opened the pull request:
A first stab at #77 

It only supports one level of nesting, which seems ideal for pseudo classes and such.  It does mean things like this will fail though:

``` css
.Component {
  &--modifier {
    .Component-element {}
}
```
Not sure if going any further sends the bem linter into `postcss-nested` territory but if a nested feature lands in the bem linter with caveats it might not be as useful. Open to thoughts on that. 

@dryoma @giuseppeg

Happy to add documentation once this is ready

giuseppeg commented:
Cool Simon! Before I even take a look at the code I was thinking that maybe we shouldn't allow modifers nesting. Modifiers should be explicit.
```css
/*
 A modifier creates something like a "context" where
 only -children, states etc are allowed.
*/
.Component--modifier { 
  &.is-state {}

  & .Component-child { /* should & be optional here? */
     &.is-active {}
     &:hover {}
     &[attr] {}
  }
}
```

This may be not trivial to implement but aligns well with the suitcss methodology imho.

giuseppeg commented on lib/validate-selectors.js:
```diff
@@ -3,6 +3,32 @@ var shouldIgnoreRule = require('./should-ignore-rule');
+}
+
+function isNestedRule(node) {
+  return /^rule$/.test(node.parent.type);
```
you could replace this with
```javascript
return node.parent.type === 'rule';
```

simonsmith replied:
True yes, that was a more complex regex before :)

giuseppeg commented on lib/validate-selectors.js:
```diff
@@ -3,6 +3,32 @@ var shouldIgnoreRule = require('./should-ignore-rule');
 var toInterpoloatedRegexp = require('./to-interpolated-regexp');
 
+function isTopLevel(node) {
+  return /^(?:atrule|root)$/.test(node.parent.type);
```
```javascript
return ['atrule', 'root'].some(function (type) { 
   return node.parent.type === type;
});
```

giuseppeg replied:
or 
```javascript
var type = node.parent.type;
return type === 'atrule' || type === 'root';
```

regexps are slow

dryoma commented:
> Modifiers should be explicit.
I agree.

As for `postcss-nested` I am a bit sceptical. stylelint authors recommend to lint code before applying any plugins. I should say, for a reason. I think it's giving way to errors added by those plugins (if they have errors). As an example, just now tried to lint this:
```
/** @define MyComponent */

.ns-MyComponent {
  display: inline; // the semicolon is lost somewhere between postcss-nested and stylelint

  &.is-open {
    display: none;
  }

  @media (max-width: 1000px) {
    display: table;
  }
}
```
with
```
...
  .pipe(gulpPostCSS([
      nested(),
      stylelint(),
      reporter({
        clearMessages: true,
        throwError: true
      })
    ], {
      //syntax: scssSyntax
    }))
```
and it falsely found an error:
```4:20    ‼  Expected a trailing semicolon (declaration-block-trailing-semicolon) [stylelint]```
If I remove the `@media` rule, the error goes away.

giuseppeg commented on test/fixtures/bem-nesting.css:
```diff
@@ -0,0 +1,20 @@
+.block-name {
+  color: green;
+
+  .block-name__element {}
```
this shouldn't be allowed, should it?

davidtheclark replied:
I also have this question.

simonsmith replied:
Do you mean it should fail the actual linter? Currently this passes fine:

``` css
.Component .Component-elem {
  color: red;
}
```

Whilst it maybe a bad practice, there are no specific rules in BEM or SUIT to say you can't prefix selectors with the component name.

davidtheclark replied:
Ok makes sense. Right.

giuseppeg commented on test/fixtures/suit-nesting.css:
```diff
@@ -0,0 +1,33 @@
+.Component {
+  color: green;
+
+  .Component-descendant {}
```
ditto, this is not valid. The nesting is only increasing the specificity.

giuseppeg commented on test/fixtures/suit-nesting.css:
```diff
@@ -0,0 +1,33 @@
+
+.Component-standAloneDescendant {}
+
+.Component {
```
this is invalid

simonsmith replied:
A bad pattern perhaps but not invalid as far as bem-linter is concerned. If we think it should be then maybe the SUIT patterns need to be adjusted?

giuseppeg replied:
Ah ok, fine with me with fixing the SUIT patterns only.
The following should be invalid in SUIT:

```css
.Component {
  .Component-descendant {}
}
```

simonsmith replied:
Cool yeah. Perhaps in strict mode we can disallow it. So:

``` css
// bad
/** @define Component */
.Component .Component-descendant {}

// good
/** @define Component; weak */
.Component .Component-descendant {}
```

We should raise a separate issue, but in the meantime I'll remove it from the test fixtures in strict mode.

simonsmith commented:
> Modifiers should be explicit.

Let me see if I understand by example:

``` css
.Component {
  color: red;

  &.is-stateName {}
  &-descendant {}
  &:pseudo-class {}
  &:pseudo-element {}
  &::pseudo-element {}
  &[attr] {}
}

/* modifier needs to start a new block? */

.Component--modifier {
  .Component-descendant {}
  &.is-stateName {}
  &:pseudo-class {}
}

/* How about modifiers on descendants? */

.Component {
  &-descendant--modifier {}
}
```

I also wonder if this whole thing should be recursive, or whether we should allow one level of nesting? It gets complex the more you think about it :)

simonsmith commented:
> As for postcss-nested I am a bit sceptical

Agreed, it's not ideal. I was just wondering if `bem-linter` should be implementing almost that whole plugin as once you support nesting users may try to put `@media` blocks in rulesets a la Sass for example.

Sounds like we would be aiming to allow nesting with specific rules though.

dryoma commented:
I get it that it's a Post**CSS**, not PostSass plugin. Still, I think that it is quite natural for a user to expect `postcss-bem-linter` to support nesting as well, since  it can be used as a plugin to `stylelint` or alongside it, and `stylelint` can support SCSS syntax via `postcss-scss` (and it's especially great it works with nesting) and doesn't need any specific preprocessing plugins for this (like `postcss-nesting` in question). Otherwise it would require a separate pipelilne for linting selectors. Not to mention the plugins may cause errors.

simonsmith commented:
@dryoma Agreed yes, but I also wonder if `bem-linter` should take this chance to lay down some rules and stop the nested hell you sometimes see in Sass and LESS

davidtheclark commented on lib/validate-selectors.js:
```diff
@@ -3,6 +3,33 @@ var shouldIgnoreRule = require('./should-ignore-rule');
+  });
+}
+
+function unWrapSelectors(parent, child) {
```
I should have suggested this: https://github.com/davidtheclark/postcss-resolve-nested-selector. Do you that module this will work for this case? We use it in stylelint for rules about nested selectors.

davidtheclark commented:
I think using `postcss-resolve-nested-selector` might solve some of the complexity concern.

I also think that people who want to restrict nesting depth could use stylelint. I think it's nice to keep this particular plugin focused solely on selectors.

simonsmith commented:
@davidtheclark Ah nice! Exactly what I needed. I'll look at implementing that and leave the depth concern to stylelint as you mentioned

dryoma commented:
> I also think that people who want to restrict nesting depth could use stylelint.

I agree, styleling does this perfectly, even with SCSS.

giuseppeg commented on lib/validate-selectors.js:
```diff
@@ -3,6 +3,33 @@ var shouldIgnoreRule = require('./should-ignore-rule');
+function unWrapSelectors(parent, child) {
+  return parent.selectors.reduce(function(selectors, parentSelector) {
+    child.selectors.forEach(function(childSelector) {
+      if (!/^&/.test(childSelector)) {
```
this could be
```javascript
if (childSelector.charAt(0) !== '&')
```

giuseppeg commented:
@simonsmith are you just unwrapping and validating the nested rules then? If so I think that the branch is ok! 

By the way are you familiar with the [CSS Nesting Module Level 3](http://tabatkins.github.io/specs/css-nesting/) proposal by Tab Atkins? Would it make sense to try to add support for this too? I mean we are introducing new rules and if we want to move away from them or remove some in the future that'd require a breaking change.

simonsmith commented:
@giuseppeg Yeah, I'm thinking do the minimum here which is just unwrapping all the selectors and validating them. Changes to those rules can happen outside this PR.

I'm currently refactoring to use `postcss-resolve-nested-selector` (via @davidtheclark) which supports [postcss-nested](https://github.com/postcss/postcss-nested/) (Sass) and [postcss-nesting](https://github.com/jonathantneal/postcss-nesting) (CSS3) so that should cover it.

Will push some changes for review shortly.

simonsmith commented:
Okay, pushed some changes.

The idea now is to detect if a rule is nested and if so traverse up to the parent and construct the selector from there. The reason for this is so we always have the correct `rule` to output as the warning if linting fails.

See what you think

simonsmith commented:
@davidtheclark Look okay to you?

davidtheclark commented on lib/validate-selectors.js:
```diff
@@ -2,6 +2,40 @@ var listSequences = require('./list-sequences');
+var resolveNestedSelector = require('postcss-resolve-nested-selector');
+
+function isTopLevel(node) {
+  return node.type === 'atrule' || node.type === 'root';
```
Are at-rules always "top level"? What if you're nesting an at-rule inside a rule, e.g.

```scss
.foo {
  &:hover { color: pink; }
  @media (min-width: 400px) {
    &:hover { color: orange; }
  }
}
```

simonsmith replied:
True, that is a valid pattern. I'll add a test and adjust accordingly.

davidtheclark commented:
Just had a couple of comments I added.

simonsmith commented:
I've pushed some more changes and added some more tests. Main change was to move the logic into a `getSelectors` function so it could be isolated in the tests. I'll post my commit message as an explanation:

```
This allows certainty that selectors are being unwrapped as expected. By
following the 'black box' style used in other tests it can lead to false
positives. For example:

.Component {
  .Component-descendant {
    color: green;
  }
}

In the above code what should be linted is
`.Component .Component-descendant` but it's quite possible due to an
error that just `.Component` is tested. If this happens it will still
pass the black box test as `.Component` is valid. This can lead to hard
to find bugs with unwrapping nesting selectors.
```

The extra tests give it a bit more confidence that this is working as expected. Have also added some CSS Nesting Module fixtures.

davidtheclark commented on lib/validate-selectors.js:
```diff
@@ -2,6 +2,56 @@ var listSequences = require('./list-sequences');
+}
+
+function hasNoDeclarations(node) {
+  return node.nodes.length && node.every(function(child) {
```
Should this be a `||`? If the node has no children, doesn't that mean it has no declarations?

simonsmith replied:
This is needed because if the array is empty then `node.every` just returns `true` which is incorrect, but if the array length was greater than `0` it still needs to be checked to ensure each item is a declaration.

simonsmith replied:
It fails if I use `||`. The main idea here is to skip rulesets that have no children at all:

``` css
.Component { 
  &--modifier {} 
}
```

Neither of these have declarations, but I'm only interested in the rule that contains another rule. The `&--modifier` ruleset still needs to be validated so I let that through.

Hope that makes sense.

davidtheclark commented:
Looks good! Just that one comment. Anybody object to merging after it's addressed?

giuseppeg commented on test/nesting.js:
```diff
@@ -0,0 +1,100 @@
+var postcss = require('postcss');
+var validateSelectors = rewire('../lib/validate-selectors');
+
+describe('getSelectors function', function() {
```
Nice to have: a test for group of selectors
```css
.Component,
.Component-descendant { 
  color: green;

  &:hover { 
      /* ... */
  } 
}
```

simonsmith replied:
Cool yeah, will add one

simonsmith replied:
Done in https://github.com/postcss/postcss-bem-linter/pull/85/commits/199ceab8a1e528d1b7b35c217272d06914c2ebe9

giuseppeg commented on test/bem-pattern.js:
```diff
@@ -7,10 +7,16 @@ describe('using BEM pattern', function() {
     var s = util.selectorTester('/** @define block */');
 
-    // mirroring tests from http://goo.gl/Db6QPu
+    // mirroring tests from https://goo.gl/G0VZHZ
```
nit: you could replace with a `git.io` shortlink

giuseppeg commented on lib/validate-selectors.js:
```diff
@@ -2,6 +2,56 @@ var listSequences = require('./list-sequences');
+  // Skip validation on rules with no declarations
+  // as these don't exist after rules have been unwrapped
+  if (hasNoDeclarations(rule)) {
+    return null;
```
In theory this could return an empty array.

simonsmith replied:
A falsey value felt more appropriate otherwise I have to check `if (!selectors.length) {return;}` further down.

giuseppeg replied:
You don't need that check at all if you return an empty array:
```javascript
[].forEach(/* ... */)
```

simonsmith replied:
That's true!

simonsmith replied:
Done in https://github.com/postcss/postcss-bem-linter/pull/85/commits/199ceab8a1e528d1b7b35c217272d06914c2ebe9

simonsmith replied:
Done here https://github.com/postcss/postcss-bem-linter/pull/85/commits/33cd25401acfc4654fedbca5d586cad2b4743234

Good catch :)

giuseppeg commented:
Looks good to me, feel free to ignore my last round of comments if you want ;)

davidtheclark commented:
Thanks all. I'm fine with merging, so ping me when you think it's ready @simonsmith.

simonsmith commented:
@davidtheclark I'm happy with this now. Commits are pretty tidy. Need anything added to the `README.md`?

davidtheclark commented:
I think it's fine. Released in 2.5.0! Thanks again everybody!

simonsmith commented:
🎉

dryoma commented:
Great, thanks @simonsmith !

marcustisater commented:
Wohoo! Great work @simonsmith! 🎉 😄

brianmcallister commented:
Little late to the party, but this doesn't appear to be working for me in v2.6.0. Here's my Sass file:

`_test.scss`
```sass
/* postcss-bem-linter: define test */
.test {
  &--red {
    color: rgba(255, 0 0);
  }

  .child {
    color: #ffffff;
  }
}
```

and here's my config:

```js
module.exports = {
  plugins: [
    'stylelint-selector-bem-pattern'
  ],
  rules: {
    'plugin/selector-bem-pattern': {
      componentName: '^\.[-a-z]+$',
      componentSelectors: '^\.{componentName}[-a-z]+$',
    },
  }
};
```

and my output:

```bash
src/stylesheets/modules/_test.scss
 7:3  ✖  Invalid component selector ".test .child"   plugin/selector-bem-pattern
```

Am I just doing something wrong? I started with a super simple regex for now, just to validate that this works.

simonsmith commented:
Hey @brianmcallister,

Do you mind adding that as a separate issue so I don't lose track of it. Thanks

brianmcallister commented:
Yep! No problem.
